### ISSUE TITLE
Another Bug in search.py (iterating over an `int` !)

### ISSUE THREAD

**Issue by goswami-rahul** [aimacode/aima-python#826]
the function `find_blank_square()` is iterating over an `int`.
In line 422. Here's the code:
```python
def find_blank_square(self, state):
        """Return the index of the blank square in a given state"""
        for row in len(state):
            for column in len(row):
                if state[row][column] == 0:
                    index_blank_square = (row, column)
        return index_blank_square
```
`len(state)` and `len(row)` will _always_ return an `int`.
I wonder how it passed the tests.

Edit: Found two more such cases!
in line 477
```python
    def goal_test(self, state):
        """Given a state, return True if state is a goal state or False, otherwise"""
        for row in len(state):
            for column in len(row):
                if state[row][col] != self.goal[row][column]:
                    return False
        return True
```
and in line 496
```python
    def h(self, state):
        """Return the heuristic value for a given state. Heuristic function used is 
        h(n) = number of misplaced tiles."""
        num_misplaced_tiles = 0
        for row in len(state):
            for column in len(row):
                if state[row][col] != self.goal[row][column]:
                    num_misplaced_tiles += 1
        return num_misplaced_tiles
```

**Comment by ad71** [comment]
This bug was introduced as a result of #733 .
This has been fixed in #807 but the PR is yet to be merged.
Many of your other issues are also already solved and the PRs await merging. I would advise you to sumbit PRs after the repository has calmed down a bit, as this might lead to merge conflicts later. It is fine if you are working on files that do not currently have open PRs. I would also implore you to have a look at the open PRs, especially those involving `search.py` and `search.ipynb`. This might solve many of your queries.

**Comment by bakerwho** [comment]
@goswami-rahul I think you should close this issue seeing how it is pretty much a more specific duplicate of #785, and has been fixed in #807.
